node/src/components/consensus/protocols/highway/synchronizer.rs

```diff
@@ -339,20 +339,24 @@ impl<I: NodeIdT, C: Context + 'static> Synchronizer<I, C> {
                 Some(pv) => pv,
             };
             if let Some(dep) = highway.missing_dependency(pv.pvv()) {
-                if self.dep_already_downloaded(&dep) {
+                if self.vertices_no_deps.contains_dependency(&dep) {
                     // `dep` is already downloaded and waiting in the synchronizer queue to be
                     // added, we don't have to request it again. Add the `pv`
-                    // back to the queue so that it can be retried later when,
-                    // `dep` is added to the protocol state.
+                    // back to the queue so that it can be retried later. `dep` does not wait for
+                    // any of the dependencies currently so it should be retried soon.
                     self.add_missing_dependency(dep.clone(), pv);
                     continue;
                 }
                 // We are still missing a dependency. Store the vertex in the map and request
                 // the dependency from the sender.
+                // Find the first dependency that `pv` needs that we haven't synchronized yet
+                // and request it from the sender of `pv`. Since it relies on it, it should have
+                // it as well.
+                let transitive_dependency = self.find_transitive_dependency(dep.clone());
```

Comment:
We should move this line to the top (above 342):
```rust
let transitive_dependency = self.find_transitive_dependency(dep.clone());
if self.vertices_no_deps.contains_dependency(&transitive_dependency) {
    self.add_missing_dependency(dep.clone(), pv);
    continue;
}
```

Because if the _transitive_ dependency is already in `vertices_no_deps`, there's also no reason to request it.

E.g.: Bob sends us `c2`, while `c1` is in `vertices_awaiting_deps`. But we already received even `c0` from Alice, and it's in `vertices_no_deps`. Then we also don't need to request `c0` from Bob.

Reply:
Good point.